File: SOC_Photon/src/myLibrary/myTables.cpp
#include <Arduino.h> //needed for Serial.println
#include "myTables.h"
#include "math.h"

// Global variables
extern char buffer[256];

// Interpolating, clipping, 1 and 2-D arbitrarily spaced table look-up

/* B I N S E A R C H
*
*   Purpose:    Find x in { v[0] <= v[1] <= ... ,= v[n-1] } and calculate
*               the fraction of range that x is positioned.
*
*   Author:     Dave Gutz 09-Jun-90.
*   Revisions:  Dave Gutz 20-Aug-90 Input x instead of *x to protect
*                   integrity of calling function.
*                      16-Aug-93   Pointers.
*   Inputs:
*       Name        Type        Length      Definition
*       x           float      1           Input to vector.
*       n           int         1           Size of vector.
*       v           float      n           Vector.
*   Outputs:
*       Name        Type        Length      Definition
*       *dx         float      1           Fraction of range for x.
*       *low        int         1           Current low end of range.
*       *high       int         1           Current high end of range.
*   Hardware dependencies:  ANSI C.
*   Header needed in scope of caller:   None.
*   Global variables used:  None.
*   Functions called:   None.
*/
void binsearch(float x, float *v, int n, int *high, int *low, float *dx)
{
  int mid;

  /* Initialize high and low  */
  *low = 0;
  *high = n - 1;

  /* Check endpoints  */
  if (x >= *(v + *high))
  {
    *low = *high;
    *dx = 0.;
  }
  else if (x <= *(v + *low))
  {
    *high = *low;
    *dx = 0.;
  }

  /* Search if necessary  */
  else
  {
    while ((*high - *low) > 1)
    {
      mid = (*low + *high) / 2;
      if (*(v + mid) > x)
        *high = mid;
      else
        *low = mid;
    }
    *dx = (x - *(v + *low)) / (*(v + *high) - *(v + *low));
  }
} /* End binsearch    */

/* T A B 1
*
*   Purpose:    Univariant arbitrarily spaced table look-up.
*
*   Author:     Dave Gutz 09-Jun-90.
*   Revisions:  Dave Gutz 20-Aug-90 Input x instead of *x to protect
*                   integrity of calling function.
*                         16-Aug-93   Pointers.
*   Inputs:
*       Name        Type        Length      Definition
*       n           int         1           Number of points.
*       x           float      1           Independent variable.
*       v           float      n           Breakpoint table.
*       y           float      n           Table data.
*   Outputs:
*       Name        Type        Length      Definition
*       tab1        float      1           Result of table lookup.
*   Hardware dependencies:  ANSI C.
*   Header needed in scope of caller:   None.
*   Global variables used:  None.
*   Functions called:   binsearch.
*/
float tab1(float x, float *v, float *y, int n)
{
  float dx;
  int high, low;
  void binsearch(float x, float *v, int n, int *high,
                 int *low, float *dx);
  if (n < 1)
    return y[0];
  binsearch(x, v, n, &high, &low, &dx);
  return *(y + low) + dx * (*(y + high) - *(y + low));
} /* End tab1 */

/* tab1clip:    Univariant arbitrarily spaced table look-up with clipping.
*
*   Author:     Dave Gutz 20-Nov-16
*   Inputs:
*       n           Number of points
*       x           Independent variable
*       v           Breakpoint table
*       y           Table data
*   Outputs:
*       tab1        Result of table lookup
*/
float tab1clip(float x, float *v, float *y, int n)
{
  float dx;
  int high, low;
  void binsearch(float x, float *v, int n, int *high,
                 int *low, float *dx);
  if (n < 1)
    return y[0];
  binsearch(x, v, n, &high, &low, &dx);
  return *(y + low) + fmax(fmin(dx, 1.), 0.) * (*(y + high) - *(y + low));
} /* End tab1clip */

/* T A B 2
*
*   Purpose:    Bivariant arbitrarily spaced table look-up.  Clips
*
*   Author:     Dave Gutz 20-Aug-90.
*   Revisions:            16-Aug-93   Pointers.
*   Inputs:
*       Name        Type        Length      Definition
*       n1          int         1           Number of ind var #1 brkpts.
*       n2          int         1           Number of ind var #2 brkpts.
*       x1          float      1           Independent variable #1.
*       x2          float      1           Independent variable #2.
*       v1          float      n1          Breakpoints for var #1.
*       v2          float      n2          Breakpoints for var #2.
*       y           float      n1*n2       Table data.
*   Outputs:
*       Name        Type        Length      Definition
*       tab2        float      1           Result of table lookup.
*   Hardware dependencies:  ANSI C.
*   Header needed in scope of caller:   None.
*   Global variables used:  None.
*   Functions called:   binsearch (natively clipping)
*/
float tab2(float x1, float x2, float *v1, float *v2, float *y, int n1,
            int n2)
{
  float dx1, dx2, r0, r1;
  int high1, high2, low1, low2, temp1, temp2;
  void binsearch(float x, float *v, int n, int *high,
                 int *low, float *dx);
  if (n1 < 1 || n2 < 1)
    return y[0];
  binsearch(x1, v1, n1, &high1, &low1, &dx1);  // clips
  binsearch(x2, v2, n2, &high2, &low2, &dx2);  // clips
  temp1 = low2 * n1 + low1;
  temp2 = high2 * n1 + low1;
  r0 = *(y + temp1) + dx1 * (*(y + low2 * n1 + high1) - *(y + temp1));
  r1 = *(y + temp2) + dx1 * (*(y + high2 * n1 + high1) - *(y + temp2));
  return r0 + dx2 * (r1 - r0);
} /* End tab2 */

// class TableInterp
// constructors
TableInterp::TableInterp()
    : n1_(0) {}
TableInterp::TableInterp(const unsigned int n, const float x[])
    : n1_(n)
{
  x_ = new float[n1_];
  for (unsigned int i = 0; i < n1_; i++)
  {
    x_[i] = x[i];
  }
}

TableInterp::~TableInterp() {}
// operators
// functions
float TableInterp::interp(void)
{
  return (-999.);
}
void TableInterp::pretty_print(void)
{
  unsigned int i;
  Serial.printf("    x={");
  for ( i = 0; i < n1_; i++ )
  {
     Serial.printf("%7.3f, ", x_[i]);
  }
  Serial.printf("};\n");
  Serial.printf("    v={");
  for ( i = 0; i < n1_; i++ )
  {
     Serial.printf("%7.3f, ", v_[i]);
  }
  Serial.printf("};\n");
}

// 1-D Interpolation Table Lookup
// constructors
TableInterp1D::TableInterp1D() : TableInterp() {}
TableInterp1D::TableInterp1D(const unsigned int n, const float x[], const float v[])
    : TableInterp(n, x)
{
  v_ = new float[n1_];
  for (unsigned int i = 0; i < n1_; i++)
  {
    v_[i] = v[i];
  }
}
TableInterp1D::~TableInterp1D() {}
// operators
// functions
float TableInterp1D::interp(const float x)
{
  return (tab1(x, x_, v_, n1_));
}

// 1-D Interpolation Table Lookup
// constructors
TableInterp1Dclip::TableInterp1Dclip() : TableInterp() {}
TableInterp1Dclip::TableInterp1Dclip(const unsigned int n, const float x[], const float v[])
    : TableInterp(n, x)
{
  v_ = new float[n1_];
  for (unsigned int i = 0; i < n1_; i++)
  {
    v_[i] = v[i];
  }
}
TableInterp1Dclip::~TableInterp1Dclip() {}
// operators
// functions
float TableInterp1Dclip::interp(const float x)
{
  return (tab1(x, x_, v_, n1_));
}

// 2-D Interpolation Table Lookup
/* Example usage:  see voc_T_ in ../Battery.cpp.
x = {x1, x2, ...xn}
y = {y1, y2, ...ym}
v = {v11, v12, ...v1n, v21, v22, ...v2n, ...............  vm1, vm2, ...vmn}
  = {v11, v12, ...v1n,
     v21, v22, ...v2n,
     ...............
     vm1, vm2, ...vmn}
*/
// constructors
TableInterp2D::TableInterp2D() : TableInterp() {}
TableInterp2D::TableInterp2D(const unsigned int n, const unsigned int m, const float x[],
                             const float y[], const float v[])
    : TableInterp(n, x)
{
  n2_ = m;
  y_ = new float[n2_];
  for (unsigned int j = 0; j < n2_; j++)
  {
    y_[j] = y[j];
  }
  v_ = new float[n1_ * n2_];
  for (unsigned int i = 0; i < n1_; i++)
    for (unsigned int j = 0; j < n2_; j++)
    {
      v_[i + j * n1_] = v[i + j * n1_];
    }
}
TableInterp2D::~TableInterp2D() {}
// operators
// functions
float TableInterp2D::interp(float x, float y)
{
  return (tab2(x, y, x_, y_, v_, n1_, n2_));  // clips
}
//tab2(float x1, float x2, float *v1, float *v2, float *y, int n1, int n2);
/*
static float  xTbl[6]  =
  {-4.7, -1.88, -1.41, -.94, -.47, 4.7};
static float   yTbl[4]   =
  {0., 10000., 20000., 30000.};
static float  vTbl[24]  =
  {5.5, 3.5, 3.0, 2.5,
   5.5, 3.5, 3.0, 2.5,
   5.5, 3.5, 3.0, 5.5,
   5.5, 3.5, 5.5, 5.5,
   5.5, 5.5, 5.5, 5.5,
   5.5, 5.5, 5.5, 5.5};
    val = tab2(xDR_ven, fxven, xTbl, yTbl, vTbl,
    sizeof(xTbl)/sizeof(float),
    sizeof(yTbl)/sizeof(float)) * 100. / 4.7;
*/

void TableInterp2D::pretty_print()
{
  unsigned int i, j;
  Serial.printf("    y={"); for ( j=0; j<n2_; j++ ) Serial.printf("%7.3f, ", y_[j]); Serial.printf("};\n");
  Serial.printf("    x={"); for ( i=0; i<n1_; i++ ) Serial.printf("%7.3f, ", x_[i]); Serial.printf("};\n");
  Serial.printf("    v={\n");
  for ( j=0; j<n2_; j++ )
  {
    Serial.printf("      {");
    for ( i=0; i<n1_; i++ ) Serial.printf("%7.3f, ", v_[j*n1_+i]);
    Serial.printf("},\n");
  }
  Serial.printf("      };\n");
}

File: SOC_Photon/py/pyDAGx/myTables.py
# Table utility
# Copyright (C) 2023 Dave Gutz
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# See http://www.fsf.org/licensing/licenses/lgpl.txt for full license text.

__author__ = 'Dave Gutz <davegutz@alum.mit.edu>'
__version__ = '$Revision: 1.1 $'
__date__ = '$Date: 2022/01/25 09:42:00 $'

import numpy as np
import Iterate


def binsearch(x_, v_, n):
    # Initialize to high and low
    low = 0
    high = n - 1
    # Check endpoints
    if x_ >= v_[high]:
        low = high
        dx = 0.
    elif x_ <= v_[low]:
        high = low
        dx = 0.
    # else search
    else:
        while high - low > 1:
            mid = int((low + high) / 2)
            if v_[mid] > x_:
                high = mid
            else:
                low = mid
        dx = (x_ - v_[low]) / (v_[high] - v_[low])
    return high, low, dx


# Check if given array is Monotonic
def isMonotonic(A):
    x_, y_ = [], []
    x_.extend(A)
    y_.extend(A)
    x_.sort()
    y_.sort(reverse=True)
    if all(np.array(x_) == np.array(A)) or all(np.array(y_) == np.array(A)):
        return True
    return False


class Error(Exception):
    """Lookup Table Error"""
    pass


class TableInterp1D:
    # Gridded lookup table, natively clipped
    """// 1-D Interpolation Table Lookup
    /* Example usage:
    x = {x1, x2, ...xn}
    v = {v1, v2, ...vn}"""

    def __init__(self, x_, v_):
        self.n = len(x_)
        self.x = x_
        if len(v_) != self.n:
            raise Error("length of array 'v' =", len(v_),
                        " inconsistent with length of 'x' = ", len(x_))
        self.v = v_

    # Interpolation
    def interp(self, x_):
        if self.n < 1:
            return self.v[0]
        high, low, dx = binsearch(x_, self.x, self.n)  # clips
        r_val = self.v[low] + dx * (self.v[high] - self.v[low])
        return float(r_val)

    def __str__(self, prefix='', spacer='  '):
        s = prefix + "(TableInterp1D):\n"
        s += spacer + "x = ["
        count = 1
        N = len(self.x)
        for X in self.x:
            s += " {:7.3f}".format(X)
            if count < N:
                s += ","
                count += 1
        s += "]\n"
        s += spacer + "t = ["
        count = 1
        N = len(self.x)
        for X in self.x:
            s += " {:7.3f}".format(self.interp(X))
            if count < N:
                s += ","
                count += 1
        s += "]\n"
        return s


class TableInterp2D:
    # Gridded lookup table, natively clipped
    """// 2-D Interpolation Table Lookup
    /* Example usage:  see voc_T_ in ../Battery.cpp.
    x = {x1, x2, ...xn}
    y = {y1, y2, ...ym}
    v = {v11, v12, ...v1n, v21, v22, ...v2n, ...............  vm1, vm2, ...vmn}
      = {v11, v12, ...v1n,
         v21, v22, ...v2n,
     ...............
         vm1, vm2, ...vmn}"""

    def __init__(self, x_, y_, v_):
        self.n = len(x_)
        self.m = len(y_)
        self.x = x_
        self.y = y_
        if len(v_) != self.n*self.m:
            raise Error("length of array 'v' =", len(v_),
                        " inconsistent with lengths of 'x' * 'y' = ", len(x_), "*", len(y_))
        self.v = v_
        self.is_monotonic = True
        for j in range(self.m):
            if not isMonotonic(self.v[j*self.n:j*self.n+self.n]):
                self.is_monotonic = False

    def interp(self, x_, y_):
        if self.n < 1 or self.m < 1:
            return self.v[0]
        high1, low1, dx1 = binsearch(x_, self.x, self.n)
        high2, low2, dx2 = binsearch(y_, self.y, self.m)
        temp1 = low2 * self.n + low1
        temp2 = high2 * self.n + low1
        r0 = self.v[temp1] + dx1 * (self.v[low2*self.n + high1] - self.v[temp1])
        r1 = self.v[temp2] + dx1 * (self.v[high2*self.n + high1] - self.v[temp2])
        return float(r0 + dx2 * (r1 - r0))

    # Reverse interpolation
    def r_interp(self, t_, y_):
        ice = Iterate("TableInterp2D.r_interp")
        SOLV_ERR = 1e-12
        SOLV_MAX_COUNTS = 25
        SOLV_SUCC_COUNTS = 4
        if self.is_monotonic:
            xmin = self.x[0]
            xmax = self.x[end]
            ice.init(xmax, xmin, 2*SOLV_ERR)  # init ice.e > SOLV_ERR
            while abs(ice.e) > SOLV_ERR and ice.count < SOLV_MAX_COUNTS and abs(ice.dx) > 0.:
                ice.increment()
                x_ = ice.x
                v_solved = self.interp(x_, y_)
                ice.e(v_solved - t_)
                ice.iterate(verbose=True, success_count=SOLV_SUCC_COUNTS, en_no_soln=True)
            return x_
        else:
            print('r_interp(myTables.py):  table not monotonic in z')
            return np.nan

    def __str__(self, prefix='', spacer='  '):
        s = prefix + "(TableInterp2D):\n"
        s += spacer + "x = ["
        count = 1
        N = len(self.x)
        for X in self.x:
            s += " {:7.3f}".format(X)
            if count < N:
                s += ","
                count += 1
        s += "]\n"
        s += spacer + "y = ["
        count = 1
        N = len(self.y)
        for Y in self.y:
            s += " {:7.3f}".format(Y)
            if count < N:
                s += ","
                count += 1
        s += "]\n"
        s += spacer + "t = ["
        countM = 1
        M = len(self.y)
        for Y in self.y:
            count = 1
            N = len(self.x)
            for X in self.x:
                s += " {:7.3f}".format(self.interp(X, Y))
                if count < N:
                    s += ","
                    count += 1
            if countM < M:
                s += ",\n"
            countM += 1
        s += "]\n"
        return s


if __name__ == '__main__':
    import sys
    import doctest
    import numpy as np
    import timeit
    from scipy.interpolate import RegularGridInterpolator as rg
    doctest.testmod(sys.modules['__main__'])

    t_x_soc = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 0.98, 1.00]
    n = len(t_x_soc)
    t_y_t = [0., 40.]
    m = len(t_y_t)
    t_voc_i = [9.0, 11.8, 12.45, 12.61, 12.8, 12.83, 12.9, 13.00, 13.07, 13.11, 13.23, 13.5,
               9.86, 12.66, 13.31, 13.47, 13.66, 13.69, 13.76, 13.86, 13.93, 13.97, 14.05, 14.4]
    t_voc_l = [[9.0, 11.8, 12.45, 12.61, 12.8, 12.83, 12.9, 13.00, 13.07, 13.11, 13.23, 13.5],
               [9.86, 12.66, 13.31, 13.47, 13.66, 13.69, 13.76, 13.86, 13.93, 13.97, 14.05, 14.4]]
    x = np.array(t_x_soc)
    y = np.array(t_y_t)
    data_interp = t_voc_i
    data_lut = np.array(t_voc_l)

    # local solution
    table2 = TableInterp2D(x, y, data_interp)

    # scipy solution
    lut_voc = rg((x, y), data_lut.T)

    print("Testing local interp")
    for y in t_y_t:
        for x in t_x_soc:
            print('x=', x, 'y=', y, 'table2=', table2.interp(x, y))
    start = timeit.default_timer()
    n = 0
    for y in np.arange(t_y_t[0], t_y_t[-1], .01):
        for x in np.arange(t_x_soc[0], t_x_soc[-1], .01):
            n += 1
            v = table2.interp(x, y)
    end = timeit.default_timer()
    t_in = (end - start)/float(n)
    print('time per call=', t_in)
    print('')

    print("Testing scipy lutt")
    for y in t_y_t:
        for x in t_x_soc:
            print('x=', x, 'y=', y, 'lut_voc=', lut_voc([x, y]))
    start = timeit.default_timer()
    n = 0
    for y in np.arange(t_y_t[0], t_y_t[-1], .01):
        for x in np.arange(t_x_soc[0], t_x_soc[-1], .01):
            n += 1
            v = lut_voc([x, y])
    end = timeit.default_timer()
    t_rg = (end - start)/float(n)
    print('time per call=', t_rg)

    print('ratio=', t_rg/t_in)
    print('')

    t_dv = [-0.9, -0.7, -0.5, -0.3, 0.0, 0.1, 0.3, 0.5, 0.7, 0.9]
    t_soc = [0, .2, .5, 1]
    t_r = [1e-6, 0.021, 0.017, 0.018, 0.006, 0.008, 0.020, 0.026, 0.040, 1e-6,
           1e-6, 1e-6, 0.016, 0.012, 0.002, 0.003, 0.006, 0.006, 1e-6, 1e-6,
           1e-6, 1e-6, 0.016, 0.012, 0.005, 0.006, 0.008, 0.010, 1e-6, 1e-6,
           1e-6, 1e-6, 1e-6, 0.018, 0.005, 0.005, 0.006, 1e-6, 1e-6, 1e-6]
    # x = np.array(t_dv)
    # y = np.array(t_soc)
    # data_interp = t_r

    # local solution
    table2 = TableInterp2D(t_dv, t_soc, t_r)
    print("Testing local interp")
    for y in t_soc:
        for x in t_dv:
            print('x=', x, 'y=', y, 'table2=', table2.interp(x, y))

